04-onion-routing.md

```diff
@@ -432,32 +434,149 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
-### Route Blinding
+## Route Blinding
+
+Nodes receiving onion packets (either onion messages, or payments) may
+hide their identity from senders by providing a `blinded_path`.  To do
+this the recipient needs to hide both the real node ids, and the
+instructions to those nodes on where to forward the onion.
 
-Nodes receiving onion packets may hide their identity from senders by
-"blinding" an arbitrary amount of hops at the end of an onion path.
+1. subtype: `blinded_path`
+2. data:
+   * [`point`:`first_node_id`]
+   * [`point`:`first_path_key`]
+   * [`byte`:`num_hops`]
+   * [`num_hops*onionmsg_hop`:`path`]
+
+1. subtype: `onionmsg_hop`
+2. data:
+    * [`point`:`blinded_node_id`]
+    * [`u16`:`enclen`]
+    * [`enclen*byte`:`encrypted_recipient_data`]
 
-When using route blinding, nodes find a route to themselves from a given
-"introduction node" and initial "blinding point". They then use ECDH with
-each node in that route to create a "blinded" node ID and an encrypted blob
-(`encrypted_data`) for each one of the blinded nodes.
+A blinded path consists of:
+1. an initial introduction point (`first_node_id`)
+2. an initial key to share a secret with the first node_id (`first_path_key`)
+3. a series of tweaked node ids (`path.blinded_node_id`)
+4. a series binary blobs encrypted to the real node ids (`path.encrypted_recipient_data`)
+   to tell them the next hop.
 
-They communicate this blinded route and the encrypted blobs to the sender.
-The sender finds a route to the introduction node and extends it with the
-blinded route provided by the recipient. The sender includes the encrypted
-blobs in the corresponding onion payloads: they allow nodes in the blinded
-part of the route to "unblind" the next node and correctly forward the packet.
+For example, Dave wants Alice to reach him via public node Bob then
+Carol.  He creates a chain of public keys ("path_keys") for Bob, Carol
+and finally himself, so he can share a secret with each of them.  These
+keys are a simple chain, so each node can derive the next `path_key` without
+having to be told explicitly.
+
+From these shared secrets, Dave creates and encrypts three `encrypted_data_tlv`s:
+1. blob_bob: For Bob to tell him to forward to Carol
+2. blob_carol: For Carol to tell her to forward to him
+3. blob_dave: For himself to indicate the path was used, and any metadata he wants.
+
+To mask the node ids, he also derives three blinding factors from the
+shared secrets, which turn Bob into Bob', Carol into Carol' and Dave
+into Dave'.
+
+So this is the `blinded_path` he hands to Alice.
+
+1. `first_node_id`: Bob
+2. `first_path_key`: the first path key for Bob 
+3. `path`: [Bob', bob_blob], [Carol', carol_blob], [Dave', dave_blob]
+
+Alice encrypts an onion to Bob', Carol', Dave' and gives it to Bob
+with the `first_path_key`.
```

Comment:
When used for payments, Alice doesn't encrypt the onion to Bob', she encrypts it to Bob and includes the `first_path_key` in the onion. That's what allows nodes between Alice and Bob to be unaware that a blinded path is used downstream. If the onion was encrypted to Bob', Bob would have no way of obtaining the `path_key` to decrypt it.

Alice uses Bob' only when she combines a blinded path from herself to Bob to the blinded path from Bob to Dave (which is what is done for onion messages).

In both cases, Bob only uses that path key to decrypt the `encrypted_data_tlv`: it's only nodes that are *not* the introduction node that use the path key to decrypt the onion.

I clarified that in https://github.com/t-bast/bolts/commit/c073ef815152e1f523c39a3ff6d7a17ebbfc8734

Reply:
I agree on the details, just not the framing.  

I think of the broad case which I'm describing here as "encrypt to the entire blinded chain", with a special exception for the payment case.  

This is spelled out clearly in the new requirements (which I just noticed did not replace `blinding` so uses the old terminology):

```
The reader of the `blinded_path`:
- MUST create its own onion to reach the `first_node_id`
- For the first entry in `path`:
  - if it is sending a payment:
    - MAY encrypt the first blinded path onion to `first_node_id` and include `blinding` as `current_blinding_point`
  - if it does not do that:
    - MUST encrypt the first blinded path onion to the first `blinded_node_id`.
    - MUST set `next_blinding_override` in the prior onion payload to `blinding`.
  - MUST include the first `path` `encrypted_recipient_data` in each onion payload within the blinded path.
```

And in the rationale:

```
Note that there are two ways for the sender to reach the introduction
point: one is to create a normal (unblinded) payment, and place the
initial blinding point in `current_path_key` along with the
`encrypted_recipient_data` in the onion payload for the introduction point to
start the blinded path. The second way (which is the only way for
onion messages) is to create a blinded path to
the introduction point, set `next_path_key_override` inside the
`encrypted_data_tlv` on the hop prior to the introduction point to the
`first_path_key`, so it is sent to the introduction node.  However, this
only works if that prior node supports blinded paths.
```